File: fp/rich/src/main/clojure/de/hhu/rich/warenkorb/domain/model/warenkorb/warenkorbzeile/Warenkorbzeile.clj
(ns de.hhu.rich.warenkorb.domain.model.warenkorb.warenkorbzeile.Warenkorbzeile
  (:require [de.hhu.rich.warenkorb.domain.model.anzahl.Anzahl :as anzahl]
            [de.hhu.rich.warenkorb.domain.model.preis.Preis :as preis]
            [de.hhu.rich.warenkorb.domain.model.warenkorb.warenkorbzeile.WarenkorbzeileID :as warenkorbzeileID])
  (:import (de.hhu.rich.warenkorb.domain.model.artikel.ArtikelID ArtikelID)))

(defprotocol IWarenkorbzeile
  (get-id [this])
  (get-artikel-id [this])
  (get-anzahl [this])
  (get-preis [this])
  (get-max-artikel-anzahl [this])
  (erhoehe-anzahl [this n])
  (reduziere-anzahl [this n])
  (berechne-gesamtpreis [this]))

(defrecord Warenkorbzeile [id artikel-id anzahl preis max-artikel-anzahl]
  IWarenkorbzeile
  (get-id [_] id)
  (get-artikel-id [_] artikel-id)
  (get-anzahl [_] anzahl)
  (get-preis [_] preis)
  (get-max-artikel-anzahl [_] max-artikel-anzahl)

  (erhoehe-anzahl [this n]
    (if (or (nil? n) (< (.anzahl n) 0))
      (throw (IllegalArgumentException. "Die Erhöhung muss größer als 0 sein"))
      (assoc this :anzahl (anzahl/erhoehe-um anzahl n))))

  (reduziere-anzahl [this n]
    (if (or (nil? n) (< (.anzahl n) 0))
      (throw (IllegalArgumentException. "Die Reduzierung muss größer als 0 sein"))
      (assoc this :anzahl (anzahl/reduziere-um anzahl n))))

  (berechne-gesamtpreis [this]
    (let [gesamt-betrag (* (:betrag (:preis this)) anzahl)]
      (preis/erstelle-preis (bigdec gesamt-betrag) (:waehrung (:preis this))))))

(defn validiere-warenkorbzeile [warenkorbzeile]
  (when (or (nil? (:id warenkorbzeile))
            (nil? (:artikel-id warenkorbzeile))
            (nil? (:anzahl warenkorbzeile)) (< (.anzahl (:anzahl warenkorbzeile)) 0)
            (nil? (:preis warenkorbzeile))
            (nil? (:max-artikel-anzahl warenkorbzeile))
            (anzahl/ist-groesser-als? (:anzahl warenkorbzeile) (:max-artikel-anzahl warenkorbzeile)))
    (throw (IllegalArgumentException. "Fehlerhafte Warenkorbzeile"))))

(defn erstelle-warenkorbzeile [artikel-id anzahl preis max-artikel-anzahl]
  (let [neue-warenkorbzeile (->Warenkorbzeile (warenkorbzeileID/erstelle-WarenkorbzeileID)
                                              (ArtikelID artikel-id)
                                              (anzahl/erstelle-anzahl anzahl)
                                              (preis/erstelle-preis preis "EUR")
                                              (anzahl/erstelle-anzahl max-artikel-anzahl))]
    (validiere-warenkorbzeile neue-warenkorbzeile)
    neue-warenkorbzeile))

File: oop/rich/src/main/java/de/hhu/rich/warenkorb/domain/model/warenkorb/warenkorbzeile/Warenkorbzeile.java
package de.hhu.rich.warenkorb.domain.model.warenkorb.warenkorbzeile;

import de.hhu.rich.warenkorb.domain.model.PflichtfeldFehlt;
import de.hhu.rich.warenkorb.domain.model.artikel.ArtikelID;
import de.hhu.rich.warenkorb.domain.model.common.anzahl.Anzahl;
import de.hhu.rich.warenkorb.domain.model.common.preis.Preis;

import java.math.BigDecimal;

public class Warenkorbzeile {
    private final WarenkorbzeileID ID;
    private final ArtikelID artikelID;
    private Anzahl anzahl;
    private final Preis preis;
    private final Anzahl maxArtikelAnzahl;

    public WarenkorbzeileID getID() {
        return ID;
    }

    public ArtikelID getArtikelID() {
        return artikelID;
    }

    public Anzahl getAnzahl() {
        return anzahl;
    }

    public Preis getPreis() {
        return preis;
    }

    public Anzahl getMaxArtikelAnzahl() {
        return maxArtikelAnzahl;
    }

    public Warenkorbzeile(WarenkorbzeileID ID, ArtikelID artikelID, Anzahl anzahl, Preis preis, Anzahl maxArtikelAnzahl) {
        this.ID = ID;
        this.artikelID = artikelID;
        this.anzahl = anzahl;
        this.preis = preis;
        this.maxArtikelAnzahl = maxArtikelAnzahl;
    }

    public void erhoeheUm(Anzahl anzahl) {
        this.anzahl = this.anzahl.erhoeheUm(anzahl);
        validiere();
    }

    public void reduziereUm(Anzahl anzahl) {
        this.anzahl = this.anzahl.reduziereUm(anzahl);
        validiere();
    }

    public Preis berechneGesamtpreis() {
        BigDecimal gesamtBetrag = getPreis().betrag().multiply(new BigDecimal(getAnzahl().anzahl()));
        return new Preis(gesamtBetrag, getPreis().waehrung());
    }

    private void validiere(){
        if (ID == null) {
            throw new PflichtfeldFehlt("WarenkorbzeileID darf nicht null sein!");
        }
        if (artikelID == null) {
            throw new PflichtfeldFehlt("ArtikelID darf nicht null sein!");
        }
        if (anzahl == null || anzahl.anzahl() < 0) {
            throw new PflichtfeldFehlt("Anzahl darf nicht null oder kleiner als 0 sein!");
        }
        if (preis == null) {
            throw new PflichtfeldFehlt("Preis darf nicht null sein!");
        }
        if (maxArtikelAnzahl == null) {
            throw new PflichtfeldFehlt("Maximale Artikel Anzahl darf nicht null sein!");
        }
        if(anzahl.istGroesserAls(maxArtikelAnzahl)) {
            throw new IllegalStateException("Maximale Anzahl von Artikel ist überschritten.");
        }
    }
}

